File: src/main/java/io/github/xxfast/design/patterns/behavioural/state/solution/State.java
package io.github.xxfast.design.patterns.behavioural.state.solution;

abstract class State {
  abstract void heatUp(Water water);
  abstract void coolDown(Water water);

  void evaporate(Water water) {
    if (!(this instanceof Liquid)) throw new IllegalStateException("Unable to evaporate matter at state " + this);
    water.state = new Gas();
  }

  void freeze(Water water) {
    if (!(this instanceof Liquid)) throw new IllegalStateException("Unable to freeze matter at state " + this);
    water.state = new Solid();
  }

  void condense(Water water) {
    if (!(this instanceof Gas)) throw new IllegalStateException("Unable to condense matter at state " + this);
    water.state = new Liquid();
  }

  void melt(Water water) {
    if (!(this instanceof Solid)) throw new IllegalStateException("Unable to melt matter at state " + this);
    water.state = new Liquid();
  }

  void sublime(Water water) {
    if (!(this instanceof Solid)) throw new IllegalStateException("Unable to sublime matter at state " + this);
    water.state = new Gas();
  }

  void depose(Water water) {
    if (!(this instanceof Gas)) throw new IllegalStateException("Unable to depose matter at state " + this);
    water.state = new Solid();
  }
}

class Solid extends State {
  @Override void heatUp(Water water) { melt(water);}
  @Override void coolDown(Water water) { }
}

class Liquid extends State {
  @Override void heatUp(Water water) {
    evaporate(water);
  }

  @Override void coolDown(Water water) {
    freeze(water);
  }
}

class Gas extends State {
  @Override void heatUp(Water water) { }
  @Override void coolDown(Water water) { condense(water); }
}

File: src/main/kotlin/io/github/xxfast/design/patterns/behavioural/state/solution/State.kt
package io.github.xxfast.design.patterns.behavioural.state.solution

sealed interface Matter {
  fun heatUp(): Matter
  fun coolDown(): Matter
}

interface Solid: Matter {
  fun melt(): Liquid
  fun sublime(): Gas
}

interface Liquid : Matter {
  fun evaporate(): Gas
  fun freeze(): Solid
}

interface Gas : Matter {
  fun depose(): Solid
  fun condense(): Liquid
}

class Ice : Solid {
  override fun melt() = Water()
  override fun sublime() = Vapour()
  override fun heatUp() = melt()
  override fun coolDown() = this
}

class Water : Liquid {
  override fun evaporate() = Vapour()
  override fun freeze() = Ice()
  override fun heatUp() = evaporate()
  override fun coolDown() = freeze()
}

class Vapour : Gas {
  override fun condense() = Water()
  override fun depose() = Ice()
  override fun heatUp() = this
  override fun coolDown() = condense()
}

val water = Water() // Liquid

fun main(){
  water.coolDown() // Solid
    .melt() // Liquid
    .heatUp() // Gas
//    .evaporate()
}
